packages/orchestration/src/examples/auto-stake-it-tap-kit.js

```diff
@@ -0,0 +1,148 @@
+import { makeTracer } from '@agoric/internal';
+import { atob } from '@endo/base64';
+import { ChainAddressShape } from '../typeGuards.js';
+
+const trace = makeTracer('AutoStakeItTap');
+
+/**
+ * @import {VTransferIBCEvent} from '@agoric/vats';
+ * @import {VowTools} from '@agoric/vow';
+ * @import {Zone} from '@agoric/zone';
+ * @import {TargetApp} from '@agoric/vats/src/bridge-target.js';
+ * @import {ChainAddress, CosmosValidatorAddress, Denom, OrchestrationAccount} from '@agoric/orchestration';
+ * @import {FungibleTokenPacketData} from '@agoric/cosmic-proto/ibc/applications/transfer/v2/packet.js';
+ */
+
+/**
+ * @typedef {{
+ *   stakingAccount: ERef<OrchestrationAccount<{ chainId: 'cosmoshub' }>>;
+ *   localAccount: ERef<OrchestrationAccount<{ chainId: 'agoric' }>>;
+ *   validator: CosmosValidatorAddress;
+ *   localChainAddress: ChainAddress;
+ *   remoteChainAddress: ChainAddress;
+ *   remoteDenom: Denom;
+ *   localDenom: Denom;
+ * }} StakingTapState
+ */
+
+const StakingTapStateShape = {
+  stakingAccount: M.remotable(),
+  localAccount: M.remotable(),
+  validator: ChainAddressShape,
+  localChainAddress: ChainAddressShape,
+  remoteChainAddress: ChainAddressShape,
+  remoteDenom: M.string(),
+  localDenom: M.string(),
+};
+harden(StakingTapStateShape);
+
+/**
+ * @param {Zone} zone
+ * @param {VowTools} vowTools
+ */
+const prepareMakeStakingTapKit = (zone, { watch }) => {
+  return zone.exoClassKit(
+    'StakingTap',
+    {
+      tap: M.interface('AutoStakeItTap', {
+        receiveUpcall: M.call(M.record()).returns(
+          M.or(VowShape, M.undefined()),
+        ),
+      }),
+      transferWatcher: M.interface('TransferWatcher', {
+        onFulfilled: M.call(M.undefined())
+          .optional(M.bigint())
+          .returns(VowShape),
+      }),
+    },
+    /** @param {StakingTapState} state */
+    state => {
+      mustMatch(state, StakingTapStateShape);
+      return /** @type {StakingTapState} */ (harden(state));
+    },
+    {
+      tap: {
+        /**
+         * Transfers from localAccount to stakingAccount, then delegates from
+         * the stakingAccount to `validator` if the expected token (remoteDenom)
+         * is received.
+         *
+         * @param {VTransferIBCEvent} event
+         */
+        receiveUpcall(event) {
+          trace('receiveUpcall', event);
+          const tx = /** @type {FungibleTokenPacketData} */ (
+            JSON.parse(atob(event.packet.data))
+          );
+          trace('receiveUpcall packet data', tx);
+
+          const { remoteDenom, localChainAddress } = this.state;
+          // ignore outgoing transfers
+          if (tx.receiver !== localChainAddress.value) {
+            return;
+          }
+          // ignore transfers that aren't for `remoteDenom`
+          if (tx.denom !== remoteDenom) {
+            return;
+          }
+          // XXX consider checking `event.packet.source_channel`
+          // and ignoring packets from unknown channels
```

Comment:
This would be easier if we had `getConnectionInfo` readily available from `orch`